bus-mapping/src/state_db.rs

```diff
@@ -12,27 +12,78 @@ lazy_static! {
     static ref CODE_HASH_ZERO: Hash = H256(keccak256(&[]));
 }
 
+/// Define any object can encode the code to a 32 bytes hash
+pub trait CodeHash: std::fmt::Debug {
+    /// encode code
+    fn hash_code(&self, code: &[u8]) -> Hash;
```

Comment:
add an `empty_code_hash` const in the trait?

Reply:
We have wrapped this trait as an dynamic trait object so it has to be [object safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety), which do not allow any const being included.

Also evm require identical empty code hash (keccak256 for `nil` ) no matter what codehash scheme has been applied. So I enforce this empty code hash inside `CodeDB` for empty code, and exporting a method for quering value of empty code hash from `CodeDB`.

Reply:
Or say the `CodeHash` trait prove the codehash for any **not empty** code